src/high-level/abstract-tensor.lisp

```diff
@@ -0,0 +1,228 @@
+    (map! (lambda (x) (* x factor)) tensor)))
+
+;; TODO: make sure this doesn't destroy the vector type
+;; TODO: Allow for any number of tensors. Might need to write something to remove keywords from rest
+(defgeneric + (source1 source2 &key target)
+  (:documentation "Add two tensors elementwise, storing results in target or creating a new tensor")
+  (:method ((tensor1 abstract-tensor) (tensor2 abstract-tensor) &key target)
+    (assert (equalp (shape tensor1) (shape tensor2))
+            () "Incompatible shapes. Cannot add tensor of shape ~a to tensor of shape ~a."
+            (shape tensor1) (shape tensor2))
+    (let ((target (or target (copy-tensor tensor1))))
+      (map-indexes
+       (shape tensor1)
+       (lambda (&rest dims)
+         (apply #'(setf tref)
+                (cl:+ (apply #'tref tensor1 dims)
+                      (apply #'tref tensor2 dims))
+                target dims)))
+      target)))
+
+(defgeneric - (source1 source2 &key target)
+  (:documentation "Subtract two tensors elementwise, storing results in target or creating a new tensor")
+  (:method ((tensor1 abstract-tensor) (tensor2 abstract-tensor) &key target)
+    (assert (equalp (shape tensor1) (shape tensor2))
+            () "Incompatible shapes. Cannot add tensor of shape ~a to tensor of shape ~a."
+            (shape tensor1) (shape tensor2))
+    (let ((target (or target (copy-tensor tensor1))))
+      (map-indexes
+       (shape tensor1)
+       (lambda (&rest dims)
+         (apply #'(setf tref)
+                (cl:- (apply #'tref tensor1 dims)
+                      (apply #'tref tensor2 dims))
+                target dims)))
+      target)))
+
+
+;; REFER to matrix-operations in quilc
+(defgeneric = (tensor1 tensor2 &optional epsilon)
+  (:documentation "Equality check for tensors")
+  (:method ((tensor1 abstract-tensor) (tensor2 abstract-tensor) &optional epsilon)
+    (declare (ignore epsilon))
+    (unless (equal (shape tensor1) (shape tensor2))
+      (return-from = nil))
+    (map-indexes
+     (shape tensor1)
+     (lambda (&rest pos)
+       (unless (equal
+                (apply #'tref tensor1 pos)
+                (apply #'tref tensor2 pos))
+         (return-from = nil))))
+    t))
+
+
+(defgeneric every (predicate tensor)
+  (:documentation "Check that every element in a tensor meets a given condition")
+  (:method ((predicate function) (tensor abstract-tensor))
+      (foreach (lambda (x)
+                 (unless (funcall predicate x)
+                   (return-from every nil)))
+               tensor)
+    t))
+
+(defgeneric some (predicate tensor)
+  (:documentation "Check that some elements in a tensor meet a given condition")
+  (:method ((predicate function) (tensor abstract-tensor))
+      (foreach (lambda (x)
+                 (when (funcall predicate x)
+                   (return-from some t)))
+               tensor)
+    nil))
+
+(defgeneric notevery (predicate tensor)
```

Comment:
Minor nit: can you add a dash between `not` and `every`?

Reply:
Why? They mimic CL's functions. Or is that your point?

Reply:
Ugh. Nevermind.

Reply:
🤷‍♂